src/Access/AuthenticationData.cpp

```diff
@@ -105,12 +106,30 @@ AuthenticationData::Digest AuthenticationData::Util::encodeBcrypt(std::string_vi
 bool AuthenticationData::Util::checkPasswordBcrypt(std::string_view password [[maybe_unused]], const Digest & password_bcrypt [[maybe_unused]])
 {
 #if USE_BCRYPT
-    int ret = bcrypt_checkpw(password.data(), reinterpret_cast<const char *>(password_bcrypt.data()));  /// NOLINT(bugprone-suspicious-stringview-data-usage)
-    /// Before 24.6 we didn't validate hashes on creation, so it could be that the stored hash is invalid
-    /// and it could not be decoded by the library
-    if (ret == -1)
-        throw Exception(ErrorCodes::AUTHENTICATION_FAILED, "Internal failure decoding Bcrypt hash");
-    return (ret == 0);
+    /// Bcrypt takes a long time to compute, so we cache the results.
+    /// To avoid storing plaintext passwords in memory we only store SHA256 of the password from the user.
+    /// We store a mapping of the pair of SHA256 of the password and bcrypt hash to the result of the comparison.
+    using SimpleCacheBase = DB::CacheBase<std::string, bool>;
+    static auto bcrypt_cache = SimpleCacheBase("LRU", CurrentMetrics::end(), CurrentMetrics::end(), /*max_size_in_bytes*/ 1024, /*max_count*/ 1024, /*size_ratio*/ 0.5);
```

Comment:
Now, why don't we use proper metrics?

Reply:
I initally thought it was overkill to add metrics here. Added them.

Contrary to its name, `max_size_in_bytes` is not actually the size of the cache in bytes, but a size of the cache in a certain units of weight, defined by WeightFunction.
https://github.com/ClickHouse/ClickHouse/blob/60e7f527f8cc2a400b4da192212b640b17ee70e6/src/Common/LRUCachePolicy.h#L11-L18

By default WeightFunction is EqualWeightFunction<Mapped>, so the weight is equal to 1 for each entry.
https://github.com/ClickHouse/ClickHouse/blob/60e7f527f8cc2a400b4da192212b640b17ee70e6/src/Common/ICachePolicy.h#L13-L20